qlty-check/src/parser/reek.rs

```diff
@@ -0,0 +1,98 @@
+use super::Parser;
+use anyhow::Result;
+use qlty_types::analysis::v1::{Category, Issue, Level, Location, Range};
+use serde::{Deserialize, Serialize};
+
+pub struct Reek {}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct ReekJson {
+  pub context: String,
+  pub lines: Vec<i32>,
+  pub message: String,
+  pub smell_type: String,
+  pub source: String,
+  pub documentation_link: String,
+}
+
+impl Parser for Reek {
+    fn parse(&self, plugin_name: &str, output: &str) -> Result<Vec<Issue>> {
+        let mut issues = vec![];
+        let reek_issues: Vec<ReekJson> = serde_json::from_str(output)?;
+
+        for smell in reek_issues {
+            let issue = Issue {
+                tool: plugin_name.into(),
+                documentation_url: smell.documentation_link,
+                message: smell.message.trim().to_string(),
+                category: Category::Lint.into(),
+                level: Level::Medium.into(),
+                rule_key: smell.smell_type,
+                location: Some(Location {
+                    path: smell.source,
+                    range: Some(Range {
+                        start_line: smell.lines[0] as u32,
```

Comment:
Are we throwing away data here? (`lines` after first element)

Reply:
I could not find an example of a multi-line smell and I tested the following scenarios:
- when the smell references a multi-line violation (too many statements e.g.)
- when the same smell type has more than a single violation in the same file

I did find it odd however that the JSON format calls the attribute `lines` and accepts an array, so maybe worth digging in more.

Reply:
OK. Aside from peeking at the Reek code it may be hard to tell.

That said, if lines is an array of line numbers, we could use the min(lines) as the start_line and max(lines) as the end line. That might be more compatible, but would be equivalent to this code if lines is a single element

This could reduce the risk of a future bug if they start using the field in a way they are not using it today. (It just makes an assumption that lines is contiguous -- however that is probably a better behavior than throwing away data).

A final option would be to adjust the parser so that if unexpected data occurs it returns an error.

Combining these, here is my off the cuff recommendation for long term reliability:

1. Add a condition to evaluate if the lines array is contiguous
2. If lines is contiguous, use the minimum as the start line and the maximum as the end line
3. If lines is more than one element and not contiguous, return an error from the parser

This may be overkill but it's probably about 10 extra lines of code and may save our butt someday.

Reply:
I did manage to look at the source code and generate examples where `lines` is an array. e.g.

```rb
# My documented class
class Foo
  def check_response
    if respond_to?(:foo)
      puts "hi"
    elsif respond_to?(:bar)
      puts "var"
    end
  end
end
```

This sees 2 instances of the `ManualDispatch` violation and generates the following (in json format):

```json
[{"context":"Foo#check_response","lines":[4,6],"message":"manually dispatches method call","smell_type":"ManualDispatch","source":"linters/reek/fixtures/basic.in.rb","documentation_link":"https://github.com/troessner/reek/blob/v6.3.0/docs/Manual-Dispatch.md"}]
```

Reply:
An great. So now the question is, how should we translate that data to Qlty formatted data most accurately

Reply:
@brynary I updated the code to handle the above use case.